### ISSUE TITLE
Shopify Access Token

### ISSUE THREAD

**Issue by Latsode** [graphql-java-generator/graphql-maven-plugin-project#107]
Hello I am making simple client that sends graphql queries to https://{myshopID}.myshopify.com/api/2021-10/graphql.json . Every request needs X-Shopify-Storefront-Access-Token: {storefront-access-token} header . I cant seem to figure out how to add such custom header for every request . I am relatively new to Spring so this might be my problem but cant seem to find any solutions . Thank you!

**Comment by Latsode** [comment]
java.lang.IllegalStateException: Failed to execute CommandLineRunner
	at org.springframework.boot.SpringApplication.callRunner(SpringApplication.java:794) ~[spring-boot-2.5.6.jar:2.5.6]
	at org.springframework.boot.SpringApplication.callRunners(SpringApplication.java:775) ~[spring-boot-2.5.6.jar:2.5.6]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:345) ~[spring-boot-2.5.6.jar:2.5.6]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1343) ~[spring-boot-2.5.6.jar:2.5.6]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1332) ~[spring-boot-2.5.6.jar:2.5.6]
	at com.client.graphql.ClientGraphqlApplication.main(ClientGraphqlApplication.java:33) ~[classes/:na]
Caused by: org.springframework.web.reactive.function.client.WebClientResponseException$Forbidden: 403 Forbidden from POST https://dvdteststore.myshopify.com/api/2021-10/graphql.json
	at org.springframework.web.reactive.function.client.WebClientResponseException.create(WebClientResponseException.java:200) ~[spring-webflux-5.3.12.jar:5.3.12]
	Suppressed: reactor.core.publisher.FluxOnAssembly$OnAssemblyException: 
Error has been observed at the following site(s):
	*__checkpoint ⇢ 403 from POST https://dvdteststore.myshopify.com/api/2021-10/graphql.json [DefaultWebClient]
Stack trace:
		at org.springframework.web.reactive.function.client.WebClientResponseException.create(WebClientResponseException.java:200) ~[spring-webflux-5.3.12.jar:5.3.12]
		at org.springframework.web.reactive.function.client.DefaultClientResponse.lambda$createException$2(DefaultClientResponse.java:213) ~[spring-webflux-5.3.12.jar:5.3.12]
		at reactor.core.publisher.FluxMap$MapSubscriber.onNext(FluxMap.java:106) ~[reactor-core-3.4.11.jar:3.4.11]
		at reactor.core.publisher.FluxOnErrorResume$ResumeSubscriber.onNext(FluxOnErrorResume.java:79) ~[reactor-core-3.4.11.jar:3.4.11]
		at reactor.core.publisher.Operators$MonoSubscriber.complete(Operators.java:1816) ~[reactor-core-3.4.11.jar:3.4.11]
		at reactor.core.publisher.FluxDefaultIfEmpty$DefaultIfEmptySubscriber.onComplete(FluxDefaultIfEmpty.java:109) ~[reactor-core-3.4.11.jar:3.4.11]
		at reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.onComplete(FluxMapFuseable.java:150) ~[reactor-core-3.4.11.jar:3.4.11]
		at reactor.core.publisher.FluxContextWrite$ContextWriteSubscriber.onComplete(FluxContextWrite.java:126) ~[reactor-core-3.4.11.jar:3.4.11]
		at reactor.core.publisher.FluxMapFuseable$MapFuseableConditionalSubscriber.onComplete(FluxMapFuseable.java:344) ~[reactor-core-3.4.11.jar:3.4.11]
		at reactor.core.publisher.FluxFilterFuseable$FilterFuseableConditionalSubscriber.onComplete(FluxFilterFuseable.java:391) ~[reactor-core-3.4.11.jar:3.4.11]
		at reactor.core.publisher.Operators$MonoSubscriber.complete(Operators.java:1817) ~[reactor-core-3.4.11.jar:3.4.11]
		at reactor.core.publisher.MonoCollect$CollectSubscriber.onComplete(MonoCollect.java:159) ~[reactor-core-3.4.11.jar:3.4.11]
		at reactor.core.publisher.FluxMap$MapSubscriber.onComplete(FluxMap.java:142) ~[reactor-core-3.4.11.jar:3.4.11]
		at reactor.core.publisher.FluxPeek$PeekSubscriber.onComplete(FluxPeek.java:260) ~[reactor-core-3.4.11.jar:3.4.11]
		at reactor.core.publisher.FluxMap$MapSubscriber.onComplete(FluxMap.java:142) ~[reactor-core-3.4.11.jar:3.4.11]
		at reactor.netty.channel.FluxReceive.onInboundComplete(FluxReceive.java:400) ~[reactor-netty-core-1.0.12.jar:1.0.12]
		at reactor.netty.channel.ChannelOperations.onInboundComplete(ChannelOperations.java:419) ~[reactor-netty-core-1.0.12.jar:1.0.12]
		at reactor.netty.channel.ChannelOperations.terminate(ChannelOperations.java:473) ~[reactor-netty-core-1.0.12.jar:1.0.12]
		at reactor.netty.http.client.HttpClientOperations.onInboundNext(HttpClientOperations.java:702) ~[reactor-netty-http-1.0.12.jar:1.0.12]
		at reactor.netty.channel.ChannelOperationsHandler.channelRead(ChannelOperationsHandler.java:93) ~[reactor-netty-core-1.0.12.jar:1.0.12]
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) ~[netty-transport-4.1.69.Final.jar:4.1.69.Final]
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) ~[netty-transport-4.1.69.Final.jar:4.1.69.Final]
		at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) ~[netty-transport-4.1.69.Final.jar:4.1.69.Final]
		at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103) ~[netty-codec-4.1.69.Final.jar:4.1.69.Final]
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) ~[netty-transport-4.1.69.Final.jar:4.1.69.Final]
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) ~[netty-transport-4.1.69.Final.jar:4.1.69.Final]
		at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) ~[netty-transport-4.1.69.Final.jar:4.1.69.Final]
		at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:436) ~[netty-transport-4.1.69.Final.jar:4.1.69.Final]
		at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:324) ~[netty-codec-4.1.69.Final.jar:4.1.69.Final]
		at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:296) ~[netty-codec-4.1.69.Final.jar:4.1.69.Final]
		at io.netty.channel.CombinedChannelDuplexHandler.channelRead(CombinedChannelDuplexHandler.java:251) ~[netty-transport-4.1.69.Final.jar:4.1.69.Final]
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) ~[netty-transport-4.1.69.Final.jar:4.1.69.Final]
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) ~[netty-transport-4.1.69.Final.jar:4.1.69.Final]
		at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) ~[netty-transport-4.1.69.Final.jar:4.1.69.Final]
		at io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1372) ~[netty-handler-4.1.69.Final.jar:4.1.69.Final]
		at io.netty.handler.ssl.SslHandler.decodeJdkCompatible(SslHandler.java:1235) ~[netty-handler-4.1.69.Final.jar:4.1.69.Final]
		at io.netty.handler.ssl.SslHandler.decode(SslHandler.java:1284) ~[netty-handler-4.1.69.Final.jar:4.1.69.Final]
		at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:507) ~[netty-codec-4.1.69.Final.jar:4.1.69.Final]
		at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:446) ~[netty-codec-4.1.69.Final.jar:4.1.69.Final]
		at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:276) ~[netty-codec-4.1.69.Final.jar:4.1.69.Final]
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) ~[netty-transport-4.1.69.Final.jar:4.1.69.Final]
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) ~[netty-transport-4.1.69.Final.jar:4.1.69.Final]
		at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) ~[netty-transport-4.1.69.Final.jar:4.1.69.Final]
		at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410) ~[netty-transport-4.1.69.Final.jar:4.1.69.Final]
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) ~[netty-transport-4.1.69.Final.jar:4.1.69.Final]
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) ~[netty-transport-4.1.69.Final.jar:4.1.69.Final]
		at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919) ~[netty-transport-4.1.69.Final.jar:4.1.69.Final]
		at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166) ~[netty-transport-4.1.69.Final.jar:4.1.69.Final]
		at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:719) ~[netty-transport-4.1.69.Final.jar:4.1.69.Final]
		at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:655) ~[netty-transport-4.1.69.Final.jar:4.1.69.Final]
		at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:581) ~[netty-transport-4.1.69.Final.jar:4.1.69.Final]
		at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493) ~[netty-transport-4.1.69.Final.jar:4.1.69.Final]
		at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:986) ~[netty-common-4.1.69.Final.jar:4.1.69.Final]
		at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) ~[netty-common-4.1.69.Final.jar:4.1.69.Final]
		at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) ~[netty-common-4.1.69.Final.jar:4.1.69.Final]
		at java.base/java.lang.Thread.run(Thread.java:834) ~[na:na]
	Suppressed: java.lang.Exception: #block terminated with an error
		at reactor.core.publisher.BlockingSingleSubscriber.blockingGet(BlockingSingleSubscriber.java:99) ~[reactor-core-3.4.11.jar:3.4.11]
		at reactor.core.publisher.Mono.block(Mono.java:1706) ~[reactor-core-3.4.11.jar:3.4.11]
		at com.graphql_java_generator.client.RequestExecutionSpringReactiveImpl.execute(RequestExecutionSpringReactiveImpl.java:166) ~[classes/:na]
		at com.client.graphql.util.QueryRootExecutor.collectionsWithBindValues(QueryRootExecutor.java:2237) ~[classes/:na]
		at com.client.graphql.util.QueryRootExecutor.collections(QueryRootExecutor.java:2173) ~[classes/:na]
		at com.client.graphql.ClientGraphqlApplication.run(ClientGraphqlApplication.java:41) ~[classes/:na]
		at org.springframework.boot.SpringApplication.callRunner(SpringApplication.java:791) ~[spring-boot-2.5.6.jar:2.5.6]
		at org.springframework.boot.SpringApplication.callRunners(SpringApplication.java:775) ~[spring-boot-2.5.6.jar:2.5.6]
		at org.springframework.boot.SpringApplication.run(SpringApplication.java:345) ~[spring-boot-2.5.6.jar:2.5.6]
		at org.springframework.boot.SpringApplication.run(SpringApplication.java:1343) ~[spring-boot-2.5.6.jar:2.5.6]
		at org.springframework.boot.SpringApplication.run(SpringApplication.java:1332) ~[spring-boot-2.5.6.jar:2.5.6]
		at com.client.graphql.ClientGraphqlApplication.main(ClientGraphqlApplication.java:33) ~[classes/:na]


Process finished with exit code 1

**Comment by etienne-sf** [comment]
Hello,

To add specific headers, you'll have to add new exchange filter functions to the Spring conf.

To do this, you can:
* Take a look at issue #105, where I provided a solution for that
* Take a look at the `com.graphql_java_generator.client.OAuthTokenExtractor` class, that builds an exchange filter that adds OAuth headers. It's located in the _graphql-java-runtime_ module.

Etienne

**Comment by Latsode** [comment]
Hello Etienne,

First of all thank you for taking your time to respond my issue . So I took a look at com.graphql_java_generator.client.OAuthTokenExtractor but it seems a little hard for me to understand everything. But I still tried my best to make it work after reading everything you provided .  This is my new ExchangeFilterFunction 
```Java
@Component
public class ShopifyHeaderExchangeFilter implements ExchangeFilterFunction {


    @Override
    public Mono<ClientResponse> filter(ClientRequest request, ExchangeFunction next) {
        ClientRequest newRequest= ClientRequest.from(request)
                .header("X-Shopify-Storefront-Access-Token","My Secret Token")
                .build();
        return next.exchange(newRequest);
    }
}
```

and I added this Bean in my Application class
```Java
@SpringBootApplication(scanBasePackageClasses = { ClientGraphqlApplication.class, GraphQLConfiguration.class, QueryRootExecutor.class })
public class ClientGraphqlApplication implements CommandLineRunner {

    @Autowired
    QueryRootExecutor queryRootExecutor;

    @Bean
    @Primary
    public WebClient webClientAllGraphQLCases(@Qualifier("graphqlEndpoint") String graphqlEndpointAllGraphQLCases, //
                                              @Autowired(required = false) @Qualifier("httpClientAllGraphQLCases") HttpClient httpClientAllGraphQLCases,
                                              @Autowired(required = false) @Qualifier("serverOAuth2AuthorizedClientExchangeFilterFunctionAllGraphQLCases") ServerOAuth2AuthorizedClientExchangeFilterFunction serverOAuth2AuthorizedClientExchangeFilterFunctionAllGraphQLCases,
                                              @Autowired ShopifyHeaderExchangeFilter shopifyHeaderExchangeFilter) {
        return GraphQLConfiguration.getWebClient(graphqlEndpointAllGraphQLCases, httpClientAllGraphQLCases,
                serverOAuth2AuthorizedClientExchangeFilterFunctionAllGraphQLCases,
				shopifyHeaderExchangeFilter);
    }
```
But I still receive same 403 Error , would you mind helping me out to further modify my ExchangeFilterFunction to make it work properly? 

Many thanks,

Latsode

**Comment by etienne-sf** [comment]
Yes, I think I can, provided that I have access to the project.

Can you publish a simple sample for this use case on github ?

And then, I'll give its link in the plugin's doc, as a sample on howto add an ExchangeFilterFunction  to the client code generated by the plugin.


To begin with, can you check that your header is actually added to the request ?
(at least, for example, by checking in debug mode that your code that adds the header is executed)

Etienne

**Comment by Latsode** [comment]
Oh yeah I can definitely do that , saw some other people having similar issues so this example might help out few people . I'll clean it up here and there and provide github repo link . And yeah it definitely worked cause I was getting 403 errors before adding exchangeFilter and now I receive response objects without any problem .

Latsode

**Comment by Latsode** [comment]
@etienne-sf  I just made public repo of my sample here https://github.com/Latsode/spring-graphql-shopify-client-sample . Would be amazing if you could have a quick look and let me know if I can improve anywhere . Do you need anything else from me to add it in plugin's doc? 

Latsode

**Comment by etienne-sf** [comment]
Thanks, as I released the 1.18 version, I'd like to reference the repo you made.

But I ran into the #109 issue.

So this one is the one I work on now

**Comment by etienne-sf** [comment]
The link is available in the client FAQ
